File: fitness_core/plates.py
"""Barbell plate loading and warmup ramp. All weights in one consistent unit.

A loadable weight is `bar + 2 * (multiset of available plates per side)`. The
greedy fill is optimal-enough for standard plate denominations.
"""

WARMUP_PCTS = [0.0, 0.40, 0.60, 0.75, 0.85, 0.93]
WARMUP_REPS = [5, 5, 3, 2, 1, 1]


def nearest_loadable(target: float, bar: float, plates: list[float]):
    """Return (achieved_weight, descending per-side plate list) nearest to target."""
    if target <= bar:
        return bar, []
    plates = sorted(plates, reverse=True)
    per_side = (target - bar) / 2
    used: list[float] = []
    rem = per_side
    for p in plates:
        while rem >= p - 1e-9:
            used.append(p)
            rem -= p
    below = bar + 2 * sum(used)
    smallest = plates[-1]
    above = below + 2 * smallest
    if abs(above - target) < abs(below - target):
        return above, sorted(used + [smallest], reverse=True)
    return below, sorted(used, reverse=True)


def warmup_ramp(opener: float, bar: float, plates: list[float]):
    """Strictly-increasing warmups from the empty bar up to ~opener.

    Returns a list of (weight, reps, per_side_plates).
    """
    out = []
    last = None
    for pct, reps in zip(WARMUP_PCTS, WARMUP_REPS):
        if pct == 0.0:
            weight, ps = bar, []
        else:
            weight, ps = nearest_loadable(opener * pct, bar, plates)
        if last is None or weight > last:
            out.append((weight, reps, ps))
            last = weight
    return out

File: packages/core/src/math/plates.ts
const WARMUP_PCTS = [0.0, 0.4, 0.6, 0.75, 0.85, 0.93];
const WARMUP_REPS = [5, 5, 3, 2, 1, 1];

export interface LoadResult {
  weight: number;
  plates: number[];
}
export interface WarmupSet {
  weight: number;
  reps: number;
  plates: number[];
}

function greedyFill(perSide: number, platesDesc: number[]): number[] {
  const used: number[] = [];
  let rem = perSide;
  for (const p of platesDesc) {
    while (rem >= p - 1e-9) {
      used.push(p);
      rem -= p;
    }
  }
  return used;
}

export function nearestLoadable(
  target: number,
  bar: number,
  plates: number[],
): LoadResult {
  if (target <= bar || plates.length === 0) return { weight: bar, plates: [] };
  const desc = [...plates].sort((a, b) => b - a);
  const used = greedyFill((target - bar) / 2, desc);
  const below = bar + 2 * used.reduce((a, b) => a + b, 0);
  const above = below + 2 * desc[desc.length - 1];
  if (Math.abs(above - target) < Math.abs(below - target)) {
    return { weight: above, plates: greedyFill((above - bar) / 2, desc) };
  }
  return { weight: below, plates: [...used].sort((a, b) => b - a) };
}

export function warmupRamp(opener: number, bar: number, plates: number[]): WarmupSet[] {
  const out: WarmupSet[] = [];
  let last: number | null = null;
  for (let i = 0; i < WARMUP_PCTS.length; i++) {
    const pct = WARMUP_PCTS[i];
    const { weight, plates: ps } =
      pct === 0.0 ? { weight: bar, plates: [] } : nearestLoadable(opener * pct, bar, plates);
    if (last === null || weight > last) {
      out.push({ weight, reps: WARMUP_REPS[i], plates: ps });
      last = weight;
    }
  }
  return out;
}
